gravitee-am-service/src/main/java/io/gravitee/am/service/impl/PasswordHistoryService.java

```diff
@@ -0,0 +1,137 @@
+
+/**
+ * Service providing password history.
+ */
+@Component
+public class PasswordHistoryService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(PasswordHistoryService.class);
+    private final PasswordHistoryRepository repository;
+    private final AuditService auditService;
+    private final PasswordEncoder passwordEncoder;
+
+    @Autowired
+    public PasswordHistoryService(@Lazy PasswordHistoryRepository repository, AuditService auditService, PasswordEncoder passwordEncoder) {
+        this.repository = repository;
+        this.auditService = auditService;
+        this.passwordEncoder = passwordEncoder;
+    }
+
+    /**
+     * Checks that a password has not already been used in the allotted history and adds it if it has not.
+     * Returns a {@link PasswordHistory} instance on completion.
+     *
+     * @param referenceType    the reference type (DOMAIN, APPLICATION)
+     * @param referenceId      id of the reference
+     * @param userId           id of user for this password
+     * @param password         the password to add
+     * @param principal        user performing this action
+     * @param passwordSettings domain/application password settings
+     * @return Single containing {@link PasswordHistory} or an error if the password was already in the history.
+     */
+    public Maybe<PasswordHistory> addPasswordToHistory(ReferenceType referenceType, String referenceId, String userId, String password, User principal, PasswordSettings passwordSettings) {
+        LOGGER.debug("Adding password history entry for user {}", userId);
+        if (passwordSettings == null || !passwordSettings.isPasswordHistoryEnabled()) {
+            return Maybe.empty();
+        }
+        return repository.findUserHistory(referenceType, referenceId, userId)
+                         .toList()
+                         .flatMap(passwordHistories -> {
+                             if (passwordAlreadyUsed(password, passwordHistories)) {
+                                 return Single.error(() -> PasswordHistoryException.passwordAlreadyInHistory(passwordSettings));
+                             }
+                             int passwordCount = passwordHistories.size();
+                             if (passwordCount >= passwordSettings.getOldPasswords()) {
+                                 passwordHistories.sort(comparing(PasswordHistory::getCreatedAt));
+                                 return repository.delete(passwordHistories.get(0).getId())
+                                                  .andThen(repository.create(getPasswordHistory(referenceType, referenceId, userId, password)));
+                             } else {
+                                 return repository.create(getPasswordHistory(referenceType, referenceId, userId, password));
+                             }
+                         })
+                         .doOnSuccess(a -> auditService.report(AuditBuilder.builder(UserAuditBuilder.class)
```

Comment:
the Domain isn't set a reference of this audit logs, so it is not registered, also maybe we should set the User in order to be able to link the user for who the passwordHistory has been created.